### PR TITLE
Add Suspicious Tick Filtering in LiveTradingDataFeed

### LINKED ISSUE #3190: Add Suspicious Live Tick Filtering
#### Expected Behavior
Suspicious ticks in live trading should not be included in `TradeBar`/`QuoteBar` aggregation.

#### Actual Behavior
Suspicious ticks in live trading are included in `TradeBar`/`QuoteBar` aggregation.

#### Potential Solution
Add suspicious tick check in the data handlers in the `LiveTradingDataFeed`

#### Reproducing the Problem
n/a

#### System Information
n/a

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
- In the `LiveTradingDataFeed` data handlers, suspicious ticks are now filtered and not included in bar aggregation

**Note**: when using `Tick` resolution, ticks will never be filtered for the algorithm, but they will be excluded from consolidator update data (when creating `TimeSlice` objects)

#### Related Issue
Closes #3190 

#### Motivation and Context
Invalid ticks were causing equity jumps in Live trading.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
New unit tests included

#### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by Doggie52** [comment]
Out of interest - what is a _suspicious_ tick? What makes a tick qualify to be suspicious?

File: Engine/DataFeeds/LiveTradingDataFeed.cs [hunk 1]
```diff
@@ -290,7 +290,7 @@ protected Subscription CreateDataSubscription(SubscriptionRequest request)
                                 else
                                 {
                                     var tick = data as Tick;
-                                    if (tick.TickType == TickType.Trade)
+                                    if (tick?.TickType == TickType.Trade && !tick.Suspicious)
```
[the remark this answers is not in the document]
Comment (mchandschuh) [reply Engine/DataFeeds/LiveTradingDataFeed.cs]: Unrelated, but perhaps providing a great argument for the existing solution, other downstream components such as [fill models](https://github.com/QuantConnect/Lean/pull/2763) would all likely want to duplicate this suspicious check, and also, and custom user models would likely forget to include this check -- so perhaps the high-level suspicious check is best... in fact... have we ever seen someone analyze the suspicious ticks?

**Comment by mchandschuh** [comment]
**TL;DR** - _and overly simplified_ - Suspicious ticks represent off-exchange executions.

@Doggie52 -- @jaredbroad can correct me if I'm wrong, but if I recall correctly, this is a value provided by the data providers. I've always viewed these ticks as trades that aren't actionable for potentially a few reasons. _I think_ the most common reason for a tick to be marked suspicious is because it was reported from a non-standard liquidity pool. This could be an OTC trade reported up to 10 seconds late per [FINRA Rule 6282](http://finra.complinet.com/en/display/display.html?rbid=2403&element_id=4356) or even an execution that filled using [dark pool liquidity](https://en.wikipedia.org/wiki/Dark_pool).

Due to the potential reporting time delay as well as the execution potentially taking place using dark pool liquidity, the data contained in suspicious ticks are often outliers and cause issues for fill models, consolidators, indicators and pricing of portfolio holdings.

File: Engine/DataFeeds/LiveTradingDataFeed.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaredbroad) [reply Engine/DataFeeds/LiveTradingDataFeed.cs]: We made the historical bars have suspicious ticks stripped out but the tick stream is there for experts who want the unfiltered version. The aim was for the live feed to mirror this, removing the bad ticks for bars but keeping them for tick data experts.

[the remark this answers is not in the document]
Comment (StefanoRaggi) [reply Engine/DataFeeds/LiveTradingDataFeed.cs]: Just to clarify, this is the current situation before this PR:

Backtesting:
- Tick resolution equity data files contain all ticks, with the last column set to 1 for bad ticks, which corresponds to the `tick.Suspicious` property set to `true`
- Lower resolution equity data files include TradeBars generated only from non-suspicious ticks
- Tick consolidators include all ticks (suspicious or not) ==> BUG, fixed in this PR

Live:
- Subscriptions at Tick resolutions pass all ticks through to the algorithm with no filter
- Subscriptions at lower resolutions include all ticks, which will be used for aggregation into bars ==> BUG, fixed in this PR
- Tick consolidators include all ticks (suspicious or not) ==> BUG, fixed in this PR

@mchandschuh The reason for the filtering in the `LiveTradingDataFeed` was basically what you wrote in the second comment (to avoid duplicating the check in many other places).

As @jaredbroad said, tick subscriptions are intentionally unfiltered, but a fix for tick consolidators was still required (in `TimeSliceFactory`).

**Comment by StefanoRaggi** [comment]
@Doggie52 I think @mchandschuh gave a good explanation of what are suspicious ticks (a.k.a bad ticks).

I found this document is a very interesting read about this problem:
https://s3-us-west-2.amazonaws.com/tick-data-s3/pdf/Tick_Data_Filtering_White_Paper.pdf

Here are a couple of recent tick charts showing a few bad ticks:

SPY - 5/13/2019

image

IBM - 5/14/2019

image

**Comment by Doggie52** [comment]
Thanks Stefano, that's really interesting!
